modules/katamari/src/main/ts/ephox/katamari/api/Optional.ts

```diff
@@ -1,141 +1,277 @@
+  /**
+   * Create a new Optional<T> that **does not** contain a value. T can be any
+   * type because we don't actually have a T.
+   */
+  public static none<T = never>(this: void): Optional<T> {
+    return Optional.singletonNone;
+  }
 
-  /** Run a function over the 'some' value.
-   *  "map" operation on the Optional functor.
+  /**
+   * Perform a transform on an Optional type. Similar to the fold method on a
+   * number of ADTs in the TinyMCE codebase, you pass two functions into this -
+   * if `this` **does not** contain a value then `onNone` will be called (with
+   * no argument) but if `this` **does** contain a value then `onSome` will be
+   * called (with the value as its argument).
+   *
+   * For the FP enthusiasts in the room, this function:
+   * 1. Could be used to implement all of the functions below
+   * 2. Forms a catamorphism
    */
-  readonly map: <U> (mapper: (x: T) => U) => Optional<U>;
+  public fold<U>(onNone: () => U, onSome: (value: T) => U): U {
+    if (this.tag) {
+      return onSome(this.value as T);
+    } else {
+      return onNone();
+    }
+  }
 
-  /** Run a side effect over the 'some' value */
-  readonly each: (worker: (x: T) => void) => void;
+  /**
+   * Determine if this Optional object contains a value.
+   */
+  public isSome(): boolean {
+    return this.tag;
+  }
 
-  /** "bind"/"flatMap" operation on the Optional Bind/Monad.
-   *  Equivalent to >>= in Haskell/PureScript; flatMap in Scala.
+  /**
+   * Determine if this Optional object **does not** contain a value.
    */
-  readonly bind: <U> (f: (x: T) => Optional<U>) => Optional<U>;
+  public isNone(): boolean {
+    return !this.tag;
+  }
 
-  /** Does this Optional contain a value that predicate? */
-  readonly exists: (f: (x: T) => boolean) => boolean;
+  // --- Functor ---
 
-  /** Do all values contained in this option match this predicate? */
-  readonly forall: (f: (x: T) => boolean) => boolean;
+  /**
+   * Perform a transform on an Optional object, **if** there is a value. If you
+   * provide a function to turn a T into a U, this is the function you use to
+   * turn an `Optional<T>` into an `Optional<U>`. If this **does** contain a
+   * value then the output will also contain a value (that value being the
+   * output of `mapper(this.value)`), and if this **does not** contain a value
+   * then neither will the output.
+   */
+  public map<U>(mapper: (value: T) => U): Optional<U> {
+    if (this.tag) {
+      return Optional.some(mapper(this.value as T));
+    } else {
+      return Optional.none();
+    }
+  }
+
+  // --- Monad ---
 
-  /** Return all values in this Optional that match the predicate.
-   *  The predicate may refine the constituent type using TypeScript type predicates.
+  /**
+   * Perform a transform on an Optional object, **if** there is a value. Unlike
+   * the map function earlier in the piece, here the transform itself also
```

Comment:
The "earlier in the piece" doesn't make sense when looking at the doc string for this alone. Each comment should be independent, not like a walkthrough where you read it in order.

- [ ] Ensure comments are independent/not relying on reading in order.